### PR TITLE
Loopflow constraint

### PR DESCRIPTION
**Pull request by pengbo-ruidian**
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

One constraint is introduced in LinearRao: max loopflow constraint.
MPVariable related to reference flow is used in this constraint; no new MPVariable is introduced.

### CODE REVIEW

File: flowbased-computation/flowbased-computation-impl/src/main/java/com/farao_community/farao/flowbased_computation/impl/LoopFlowComputation.java [hunk 1]
```diff
@@ -101,38 +108,40 @@ private static void buildFlowFromNetwork(Network network, Crac crac, Map<Cnec, D
...
             refNpCountry.put(country, countryNetPositionValue);
         }
-
         return refNpCountry;
     }
 
-    public Map<String, Double> buildLoopFlowsFromResult(Map<Cnec, Double> frefResults, Map<Cnec, Map<String, Double>> ptdfResults, Map<String, Double> referenceNetPositionByCountry) {
-        //calculate equation 10 and equation 11 in Article 17
-        Map<String, Double> fzeroNpResults = new HashMap<>();
+    public Map<Cnec, Double> buildLoopFlowShift(Map<Cnec, Map<String, Double>> ptdfResults, Map<String, Double> referenceNetPositionByCountry) {
```
[the remark this answers is not in the document]
Comment (pengbo-ruidian) [reply flowbased-computation/flowbased-computation-impl/src/main/java/com/farao_community/farao/flowbased_computation/impl/LoopFlowComputation.java]: OK

File: data/crac/crac-loopflow-extension/src/main/java/com/farao_community/farao/data/crac_loopflow_extension/CracLoopFlowExtension.java [hunk 2]
```diff
@@ -22,6 +23,9 @@
     private GlskProvider glskProvider; //glsk for loop flow
     private List<String> countriesForLoopFlow; // list of countries for loop flow
 
+    private Map<Cnec, Map<String, Double>> ptdfs; //memorize previously calculated ptdf, set from LoopFlowComputationResult
+    private Map<String, Double> netPositions; //memorize previously calculated net postions, set from LoopFlowComputationResult
```
[the remark this answers is not in the document]
Comment (pengbo-ruidian) [reply data/crac/crac-loopflow-extension/src/main/java/com/farao_community/farao/data/crac_loopflow_extension/CracLoopFlowExtension.java]: removed

File: flowbased-computation/flowbased-computation-impl/src/main/java/com/farao_community/farao/flowbased_computation/impl/LoopFlowComputation.java [hunk 3]
```diff
@@ -34,26 +35,32 @@
...
-        Map<Cnec, Map<String, Double>> ptdfResults = computePtdfOnCurrentNetwork(); // get ptdf
-        Map<String, Double> referenceNetPositionByCountry = getRefNetPositionByCountry(network, countries); // get Net positions
-        return buildLoopFlowsFromResult(frefResults, ptdfResults, referenceNetPositionByCountry);
+    public LoopFlowComputationResult calculateLoopFlows(Network network) {
+        Map<Cnec, Double> frefResults = computeRefFlowOnCurrentNetwork(network); //get reference flow
+        Map<Cnec, Map<String, Double>> ptdfResults = computePtdfOnCurrentNetwork(network); // get ptdf
+        Map<String, Double> referenceNetPositionByCountry = getRefNetPositionByCountry(network); // get Net positions
+        Map<Cnec, Double> loopFlowShifts = buildLoopFlowShift(ptdfResults, referenceNetPositionByCountry); //compute PTDF * NetPosition
+        Map<String, Double> loopflows = buildLoopFlowsFromResult(frefResults, loopFlowShifts); //compute loopflow
+        return new LoopFlowComputationResult(ptdfResults, referenceNetPositionByCountry, loopFlowShifts, loopflows);
```
[the remark this answers is not in the document]
Comment (pengbo-ruidian) [reply flowbased-computation/flowbased-computation-impl/src/main/java/com/farao_community/farao/flowbased_computation/impl/LoopFlowComputation.java]: OK

File: ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/LinearRao.java [hunk 4]
```diff
@@ -80,7 +82,12 @@ public String getVersion() {
 
         SystematicSensitivityAnalysisResult tempSensitivityAnalysisResult;
 
-        LinearRaoModeller linearRaoModeller = createLinearRaoModeller(crac, network, preOptimSensitivityAnalysisResult);
+        //flag for using loopflow constraint, pass to LinearRaoModeller
+        SearchTreeRaoParameters searchTreeRaoParameters = parameters.getExtension(SearchTreeRaoParameters.class);
+        CracLoopFlowExtension cracLoopFlowExtension = crac.getExtension(CracLoopFlowExtension.class);
+        boolean useLoopFlow = !Objects.isNull(searchTreeRaoParameters) && searchTreeRaoParameters.isRaoWithLoopFlow() && !Objects.isNull(cracLoopFlowExtension);
+
+        LinearRaoModeller linearRaoModeller = createLinearRaoModeller(crac, network, preOptimSensitivityAnalysisResult, useLoopFlow);
```
[the remark this answers is not in the document]
Comment (pengbo-ruidian) [reply ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/LinearRao.java]: use loop flow is a config setting in Search Tree Rao Parameter currently, we are now in LinearRao
- we can add a setting of "use loop flow" also in LinearRao Parameter? but we need to make sure that both SearchTreeRao and LinearRao has the same value of "use loop flow";
- or we can move it up to RaoParameter? (this was a suggestion of Philippe in last PR)

File: ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/fillers/MaxLoopFlowFiller.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    /**
+     * update max loopflow constraint's bounds; new bounds are calculated following network update
+     * Note: currently we only focus on preventive state cnec, and we use the ptdf in preventive state. In the future,
+     * we can also update the ptdf and netposition during each iteration. In this case, instead of get ptdf directly
+     * from cracLoopFlowExtension, we need to get GlskProvider and List countries from cracLoopFlowExtension, and then
+     * launch a LoopFlowComputation to get an updated ptdfs.
+     */
+    private void updateMaxLoopFlowConstraint() {
```
[the remark this answers is not in the document]
Comment (pengbo-ruidian) [reply ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/fillers/MaxLoopFlowFiller.java]: update() empty.

File: data/crac/crac-loopflow-extension/src/main/java/com/farao_community/farao/data/crac_loopflow_extension/CracLoopFlowExtension.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (pengbo-ruidian) [reply data/crac/crac-loopflow-extension/src/main/java/com/farao_community/farao/data/crac_loopflow_extension/CracLoopFlowExtension.java]: > Actually, I'm thinking that we have the same issue with the CnecLoopFlowExtensions

Should we remove also the one in CnecLoopFlowExtensions also?

File: ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/LinearRao.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (phiedw) [reply ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/LinearRao.java]: I would move it up to RaoParameter. I feel it's the kind of parameter that does not depend on what algorithm you want to use to optimize the remedial actions.

[the remark this answers is not in the document]
Comment (phiedw) [reply ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/LinearRao.java]: (I agree with baptiste though, I feel like the LinearRao class is already doing a lot, and just passing the parameters down to the modeller rather computing the useLoopFlow boolean in the LinearRao will help)

File: ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/fillers/MaxLoopFlowFiller.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+     * Loopflow F_(0,all)_current = flowVariable - PTDF * NetPosition, where flowVariable is a MPVariable in linearRao
+     * then max loopflow MPConstraint is: - maxLoopFlow <= currentLoopFlow <= maxLoopFlow
+     * we define loopFlowShift = PTDF * NetPosition, then
+     * -maxLoopFlow + loopFlowShift <= flowVariable <= maxLoopFlow + loopFlowShift,
+     */
+    private void buildMaxLoopFlowConstraint() {
+        LoopFlowComputation loopFlowComputation = new LoopFlowComputation(linearRaoData.getCrac(), cracLoopFlowExtension);
+        Map<Cnec, Map<String, Double>> ptdfResults = loopFlowComputation.computePtdfOnCurrentNetwork(linearRaoData.getNetwork()); // get ptdf
+        Map<String, Double> referenceNetPositionByCountry = loopFlowComputation.getRefNetPositionByCountry(linearRaoData.getNetwork()); // get Net positions
+        Map<Cnec, Double> loopFlowShifts = loopFlowComputation.buildZeroBalanceFlowShift(ptdfResults, referenceNetPositionByCountry); //compute PTDF * NetPosition
```
Comment (phiedw) [inline ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/fillers/MaxLoopFlowFiller.java]: I would refactor the LoopFlowComputation class to be able to directly call the buildZeroBalanceFlowShift with a network. (You could simply overload the current buildZeroBalanceFlowShift method and put these few steps in it)

File: ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/LinearRao.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (pengbo-ruidian) [reply ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/LinearRao.java]: Thanks, modified as suggested

File: ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/fillers/MaxLoopFlowFiller.java (hunk 6 above)
Comment (pengbo-ruidian) [reply ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/fillers/MaxLoopFlowFiller.java]: OK

File: ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/LinearRao.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (pengbo-ruidian) [reply ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/LinearRao.java]: OK

File: data/crac/crac-loopflow-extension/src/main/java/com/farao_community/farao/data/crac_loopflow_extension/CracLoopFlowExtension.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (pengbo-ruidian) [reply data/crac/crac-loopflow-extension/src/main/java/com/farao_community/farao/data/crac_loopflow_extension/CracLoopFlowExtension.java]: The CracLoopFlowExtension is cleaned.
We leave the CnecLoopFlowExtension problem for the future along with refactoring of ResultVariantManager?

[the remark this answers is not in the document]
Comment (baptsegu) [reply data/crac/crac-loopflow-extension/src/main/java/com/farao_community/farao/data/crac_loopflow_extension/CracLoopFlowExtension.java]: > We leave the CnecLoopFlowExtension problem for the future along with refactoring of ResultVariantManager?

As discussed this afternoon : YES ;)

File: ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/fillers/MaxLoopFlowFiller.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+import com.farao_community.farao.linear_rao.LinearRaoProblem;
+import com.google.ortools.linearsolver.MPConstraint;
+import com.google.ortools.linearsolver.MPVariable;
+
+import java.util.*;
+
+/**
+ * Filler of loopflow constraint in linear rao problem.
+ * - Current loopflow will only be checked for preventive state cnec.
+ * - This constraint is set at beginning of search tree rao. It is not updated during optimization. It could be updated by
```
Comment (baptsegu) [inline ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/fillers/MaxLoopFlowFiller.java]: at the beginning of the linear rao

File: distribution/src/main/resources/config.yml [hunk 8]
```diff
@@ -64,11 +64,11 @@ rao:
 rao-parameters:
     dc-mode: false
     ac-to-dc-fallback: false
+    rao-with-loop-flow: false
```
Comment (baptsegu) [inline distribution/src/main/resources/config.yml]: maybe something a bit more specific : rao-with-loop-flow-limitation

File: ra-optimisation/rao-api/src/main/java/com/farao_community/farao/rao_api/RaoParameters.java [hunk 9]
```diff
@@ -69,6 +69,7 @@ protected static void load(RaoParameters parameters, PlatformConfig platformConf
             .ifPresent(config -> {
                 parameters.setDcMode(config.getBooleanProperty("dc-mode", DEFAULT_DC_MODE));
                 parameters.setAcToDcFallback(config.getBooleanProperty("ac-to-dc-fallback", DEFAULT_AC_TO_DC_FALLBACK));
+                parameters.setRaoWithLoopFlow(config.getBooleanProperty("rao-with-loop-flow", DEFAULT_RAO_WITH_LOOP_FLOW));
```
Comment (baptsegu) [inline ra-optimisation/rao-api/src/main/java/com/farao_community/farao/rao_api/RaoParameters.java]: same, rao-with-loop-flow-limitation

File: distribution/src/main/resources/config.yml (hunk 8 above)
Comment (pengbo-ruidian) [reply distribution/src/main/resources/config.yml]: ok

File: ra-optimisation/rao-api/src/main/java/com/farao_community/farao/rao_api/RaoParameters.java (hunk 9 above)
Comment (pengbo-ruidian) [reply ra-optimisation/rao-api/src/main/java/com/farao_community/farao/rao_api/RaoParameters.java]: ok

File: ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/fillers/MaxLoopFlowFiller.java (hunk 7 above)
Comment (pengbo-ruidian) [reply ra-optimisation/linear-rao/src/main/java/com/farao_community/farao/linear_rao/fillers/MaxLoopFlowFiller.java]: ok

**Comment by pengbo-ruidian** [comment]
The remaining problem is to deal with ResultVariantManager for CnecLoopFlowExtension

File: flowbased-computation/flowbased-computation-impl/src/main/java/com/farao_community/farao/flowbased_computation/impl/LoopFlowComputation.java [hunk 10]
```diff
@@ -101,26 +110,26 @@ private static void buildFlowFromNetwork(Network network, Crac crac, Map<Cnec, D
...
             Cnec cnec = entry.getKey();
-            Map<String, Double> cnecptdf = entry.getValue();
+            Map<Country, Double> cnecptdf = entry.getValue();
             double sum = 0.0;
             // calculate PTDF * NP(ref)
-            for (Map.Entry<String, Double> e : cnecptdf.entrySet()) {
-                String country = e.getKey();
+            for (Map.Entry<Country, Double> e : cnecptdf.entrySet()) {
+                Country country = e.getKey();
                 sum += cnecptdf.get(country) * referenceNetPositionByCountry.get(country);
```
Comment (baptsegu) [inline flowbased-computation/flowbased-computation-impl/src/main/java/com/farao_community/farao/flowbased_computation/impl/LoopFlowComputation.java]: This is the reason why I added this fix : here the keys where different : 
* GLSK (cnecptdf) : key (String) was "EICode + timestamp".
* refNetPosition : key(String) was "countrycode"

Comment (baptsegu) [reply flowbased-computation/flowbased-computation-impl/src/main/java/com/farao_community/farao/flowbased_computation/impl/LoopFlowComputation.java]: Now both keys are Country to make comparison more robust

Comment (pengbo-ruidian) [reply flowbased-computation/flowbased-computation-impl/src/main/java/com/farao_community/farao/flowbased_computation/impl/LoopFlowComputation.java]: All good for me.

**Pull request merged**